### PR TITLE
Allow aggregation functions in expression context

### PR DESCRIPTION
**Pull request by mattnibs**
Allow the use of aggregation functions in the expression context. These
so called "streaming aggregations" will output the current result of the
stateful aggregation each time a new value is added to it.

Part of #3018

### CODE REVIEW

**Comment by mattnibs** [comment]
Hold off on review this, I have to add one more feature.

File: compiler/kernel/expr.go [hunk 1]
```diff
@@ -100,11 +100,32 @@ func compileExpr(zctx *zed.Context, scope *Scope, e dag.Expr) (expr.Evaluator, e
...
 	case *dag.MapExpr:
 		return compileMapExpr(zctx, scope, e)
+	case *dag.Agg:
+		return compileAggExpr(zctx, scope, e)
 	default:
 		return nil, fmt.Errorf("invalid expression type %T", e)
 	}
 }
 
+func compileAggExpr(zctx *zed.Context, scope *Scope, e *dag.Agg) (expr.Evaluator, error) {
```
Comment (nwt) [inline compiler/kernel/expr.go:110]: Nits: Call this compileAgg since it compiles a dag.Agg, not a dag.AggExpr. And rename e to a for a similar reason.

File: compiler/parser/parser.peg [hunk 2]
```diff
@@ -443,7 +443,8 @@ SampleProc
           "keys": ARRAY(MAP("kind": "Assignment",
                    "lhs": MAP("kind": "ID", "name": "shape"),
                    "rhs": MAP("kind": "Call", "name": "typeof",
-                              "args": ARRAY(e)))),
+                              "args": ARRAY(e),
+                              "where": NULL))),
```
Comment (nwt) [inline compiler/parser/parser.peg:447]: Should we add `"where": NULL` to all of the `"kind": "Call"` maps? (Honest question because I forget what our philosophy is here regarding NULL map values.)

Comment (mattnibs) [reply compiler/parser/parser.peg:447]: Good point let me double check, though I'm pretty sure I wouldn't have added this if this wasn't causing something to fail.

Comment (mattnibs) [reply compiler/parser/parser.peg:447]: This was needed because this was causing our ztests to fail that was comparing the raw results of pigeon to the raw results of ast.Proc. I fixed this by having ast.Proc omit the where clause in ast.Call if where is empty. I think this is a better route.

File: compiler/parser/parser.peg [hunk 3]
```diff
@@ -725,7 +725,10 @@ Cast
     }
 
 Function
-  = !FuncGuard fn:IdentifierName __ "(" __ args:OptionalExprs __ ")" {
+  = !FuncGuard fn:IdentifierName __ "(" __ args:OptionalExprs __ ")" where:WhereClause {
+      RETURN(MAP("kind": "Call", "name": fn, "args": args, "where": where))
+    }
+  / !FuncGuard fn:IdentifierName __ "(" __ args:OptionalExprs __ ")" {
```
Comment (nwt) [inline compiler/parser/parser.peg:731]: I think we should be consistent with the way we've handled this elsewhere, which would be to add
```
  = !FuncGuard fn:IdentifierName __ "(" __ args:OptionalExprs __ ")" where:WhereClause? {
      RETURN(MAP("kind": "Call", "name": fn, "args": args, "where": where))
    }
```
here and then add `"where": NULL` to the other `"kind": "Call"` maps.

File: compiler/ztests/where-on-func.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+script: |
+  ! zq 'cut hex := to_hex(this) where this % 2 == 0' -
```
Comment (nwt) [inline compiler/ztests/where-on-func.yaml:2]: You could make this a Zed-style test.

File: compiler/parser/parser.peg (hunk 3 above)
Comment (mattnibs) [reply compiler/parser/parser.peg:731]: Ok!

File: compiler/ztests/where-on-func.yaml (hunk 4 above)
Comment (mattnibs) [reply compiler/ztests/where-on-func.yaml:2]: zed-style tests report errors?

Comment (nwt) [reply compiler/ztests/where-on-func.yaml:2]: Yes, with `errorRE: ...`.

File: compiler/parser/parser.peg (hunk 3 above)
Comment (mattnibs) [reply compiler/parser/parser.peg:731]: Back to the way it was!

File: expr/agg.go [hunk 5]
```diff
@@ -59,3 +59,27 @@ func (a *Aggregator) filter(rec *zed.Value) bool {
 	}
 	return !a.where(rec)
 }
+
+// AggregatorExpr is an expression variant of an Aggregator also referred to as
+// a streaming aggregation. They are stateful like Aggregations in a 'reduce'
+// context, but they return their current state after every input is consumed.
+func NewAggregatorExpr(agg *Aggregator) Evaluator {
```
Comment (nwt) [inline expr/agg.go:66]: Reword this to describe NewAggregatorExpr or else s/AggregatorExpr/aggregatorExpr/ and then move the comment to that type.

File: compiler/parser/parser.peg [hunk 6]
```diff
@@ -445,7 +445,8 @@ SampleProc
           "keys": ARRAY(MAP("kind": "Assignment",
                    "lhs": MAP("kind": "ID", "name": "shape"),
                    "rhs": MAP("kind": "Call", "name": "typeof",
-                              "args": ARRAY(e)))),
+                              "args": ARRAY(e),
+                              "where": NULL))),
```
Comment (nwt) [inline compiler/parser/parser.peg:449]: Maybe add `"where": NULL` to the `"kind": "Call"` maps up in the `TBD` rule to avoid surprises should someone resurrect that rule?

**Review by nwt — APPROVED** [review]

File: compiler/parser/parser.peg (hunk 6 above)
Comment (mattnibs) [reply compiler/parser/parser.peg:451]: Why does that `TBD` rule still exist?
